Simple hardware encoding

NodudeWasTaken opened the pull request:
## What
This commit test a bunch of common hardware accelerated codec's at boot.
If the user has then enabled hardware encoding in the UI, and hw acceleration is available, it replaces the mp4/h264 and webm/vp9 transcoding with available hardware accelerated codecs.
## Why
Encoding is alot more expensive than decoding, which is why it makes sense to use hw encoding when available.
Regarding the specific implementation:
It would be meaningless to try unavailable codecs, so we should remember which codecs work.
It doesnt add full hardware transcoding, simply because the requirements are too strict and it would require alot of checks to verify it works.
## Working encoders
NVidia NVENC H.264 is tested and working.
Intel QSV H.264 is tested and working.
Intel and VAAPI VP9 is assumed to work.
VAAPI H.264 doesn't work for direct file transcoding, but is enabled for HLS.
h264_v4l2m2m doesn't work for direct file transcoding, but is enabled for HLS.
## Not working/disabled encoders
The rest are disabled as i have no way of testing them.
## Docker image
I also created an additional docker image built on NVidia CUDA ubuntu, since CUDA is needed for GPU passthrough to cuda, and alpine is incompatible with NVidia drivers

bnkai commented on pkg/ffmpeg/codec_detector.go:
```diff
@@ -0,0 +1,80 @@
+		}
+	}
+
+	fmt.Println("Supported HW codecs:")
+	for _, codec := range HWCodecSupport {
+		fmt.Println("\t", codec.codec)
+	}
```
instead of `fmt` stash's logger used be used here?

NodudeWasTaken replied:
I have pushed a commit which uses the stash logger

bnkai commented on pkg/ffmpeg/codec_detector.go:
```diff
@@ -0,0 +1,80 @@
+		*/
+	} {
+		var args Args
+		args = append(args, "-hide_banner")
```
For vaapi hw accelarated encoding an extra step is needed to select the device (for most/simple cases /dev/dri/renderD128 should work).
Also one more filter to convert to a vaapi supported format and upload to the device
For example a minimal working ffmpeg command (usig an old Vega56 amd cardf would be 
```
ffmpeg -vaapi_device /dev/dri/renderD128 -f lavfi -i color=c=red -t 1 -vf format=nv12,hwupload -c:v h264_vaapi test.mp4
```
The above should work for detection purposes, for the encoding i am not sure if the issue mentioned here <https://github.com/stashapp/stash/pull/1041> (amd specific) is still valid. I will post again after testing properly.

NodudeWasTaken replied:
I have pushed a commit which preps appropriate hw devices such as vaapi_device, it also attempts to use hw scaling.
I can't test AMD, but it works for NVidia and Intel.
It also contains code for full hardware transcoding, which remains unused since i dont have a good strategy for how i should use it.

bnkai replied:
After testing a bit ( Vega56 Card using VAAPI via the opensource mesa driver ) it seems that the issue in #1041  ( empty_moov movflag option for h264_vaapi) is still present. Not sure if it is specific to my card or for all amd cards.
Given the above i couldnt test properly. A few issues (VAAPI related) with the latest commits

1.  The way HWCodecPrepend_Encode is used you end up with two different filter sections when adding the scale filter
2. In HWCodecFilter appending the `hwdownload` is not needed and breaks the encoding. Also due to 1 you already have a video filter with the format and hwupload so the end result is also broken
3. The StreamFormatV264 needed some tweaking

The closest i get it to work was with the below
```diff
diff --git a/pkg/ffmpeg/codec_detector.go b/pkg/ffmpeg/codec_detector.go
index a2cc63e8..c6c129db 100644
--- a/pkg/ffmpeg/codec_detector.go
+++ b/pkg/ffmpeg/codec_detector.go
@@ -14,6 +14,7 @@ func FindHWCodecs(ctx context.Context, encoder FFMpeg) {
 	for _, codec := range []StreamFormat{
 		StreamFormatN264,
 		StreamFormatI264,
+		StreamFormatV264,
 		/*
 			Untested:
 				StreamFormatA264,
@@ -32,14 +33,12 @@ func FindHWCodecs(ctx context.Context, encoder FFMpeg) {
 		args = args.Input("color=c=red")
 		args = args.Duration(0.1)
 
-		args = HWCodecPrepend_Encode(args, codec.codec)
 		args = args.VideoCodec(codec.codec)
 		if len(codec.extraArgs) > 0 {
 			args = append(args, codec.extraArgs...)
 		}
-
+		videoFilter := HWFilterInit(codec.codec)
 		// Test scaling
-		var videoFilter VideoFilter
 		videoFilter = videoFilter.ScaleDimensions(-2, 160)
 		videoFilter = HWCodecFilter(videoFilter, codec.codec)
 		args = args.VideoFilter(videoFilter)
@@ -93,7 +92,7 @@ func HWCodecVideoSupported(ctx context.Context, encoder FFMpeg, o TranscodeStrea
 	args = args.Duration(0.1)
 
 	// Test scaling
-	var videoFilter VideoFilter
+	videoFilter := HWFilterInit(o.Codec.codec)
 	videoFilter = videoFilter.ScaleDimensions(-2, 160)
 	videoFilter = HWCodecFilter(videoFilter, o.Codec.codec)
 	args = args.VideoFilter(videoFilter)
@@ -120,23 +119,24 @@ func HWCodecDevice_Encode(args Args, codec VideoCodec) Args {
 		args = append(args, "0")
 	case VideoCodecV264,
 		VideoCodecVVP9:
-		args = append(args, "-hwaccel_device")
+		args = append(args, "-vaapi_device")
 		args = append(args, "/dev/dri/renderD128")
 	}
 
 	return args
 }
 
-// Prepend codec for hardware encoding only
-func HWCodecPrepend_Encode(args Args, codec VideoCodec) Args {
+// Initialise a video filter for HW encoding
+func HWFilterInit(codec VideoCodec) VideoFilter {
+	var videoFilter VideoFilter
 	switch codec {
 	case VideoCodecV264,
 		VideoCodecVVP9:
-		args = append(args, "-vf")
-		args = append(args, "format=nv12,hwupload")
+		videoFilter = videoFilter.Append("format=nv12")
+		videoFilter = videoFilter.Append("hwupload")
 	}
 
-	return args
+	return videoFilter
 }
 
 /*
@@ -156,11 +156,7 @@ func HWCodecDevice_Full(args Args, codec VideoCodec) Args {
 		args = append(args, "0")
 	case VideoCodecV264,
 		VideoCodecVVP9:
-		args = append(args, "-hwaccel")
-		args = append(args, "vaapi")
-		args = append(args, "-hwaccel_output_format")
-		args = append(args, "vaapi")
-		args = append(args, "-hwaccel_device")
+		args = append(args, "-vaapi_device")
 		args = append(args, "/dev/dri/renderD128")
 	case VideoCodecI264,
 		VideoCodecIVP9:
@@ -183,7 +179,7 @@ func HWCodecFilter(args VideoFilter, codec VideoCodec) VideoFilter {
 			args = VideoFilter(strings.Replace(sargs, "scale=", "hwupload_cuda,scale_cuda=", 1)).Append("hwdownload")
 		case VideoCodecV264,
 			VideoCodecVVP9:
-			args = VideoFilter(strings.Replace(sargs, "scale=", "hwupload,scale_vaapi=", 1)).Append("hwdownload")
+			args = VideoFilter(strings.Replace(sargs, "scale=", "scale_vaapi=", 1))
 			// BUG: scale_qsv is seemingly broken on windows?
 			/*case VideoCodecI264,
 			VideoCodecIVP9:
diff --git a/pkg/ffmpeg/stream.go b/pkg/ffmpeg/stream.go
index 24d754ef..c6d0b04b 100644
--- a/pkg/ffmpeg/stream.go
+++ b/pkg/ffmpeg/stream.go
@@ -140,8 +140,7 @@ var (
 		MimeType: MimeMp4,
 		extraArgs: []string{
 			"-movflags", "frag_keyframe+empty_moov",
-			"-quality", "50",
-			"-pix_fmt", "yuv420p",
+			"-qp", "20",
 		},
 	}
 
@@ -153,7 +152,7 @@ var (
 		format:   FormatMP4,
 		MimeType: MimeMp4,
 		extraArgs: []string{
-			"-movflags", "frag_keyframe",
+			"-movflags", "frag_keyframe+empty_moov",
 			"-pix_fmt", "yuv420p",
 		},
 	}
@@ -289,7 +288,7 @@ func (o TranscodeStreamOptions) getStreamArgs() Args {
 		args = args.Duration(hlsSegmentLength)
 	}
 
-	args = HWCodecPrepend_Encode(args, o.Codec.codec)
+	videoFilter := HWFilterInit(o.Codec.codec)
 	args = args.Input(o.Input)
 
 	if o.VideoOnly {
@@ -300,7 +299,6 @@ func (o TranscodeStreamOptions) getStreamArgs() Args {
 
 	// don't set scale when copying video stream
 	if o.Codec.codec != VideoCodecCopy {
-		var videoFilter VideoFilter
 		videoFilter = videoFilter.ScaleMax(o.VideoWidth, o.VideoHeight, o.MaxTranscodeSize)
 		videoFilter = HWCodecFilter(videoFilter, o.Codec.codec)
 		args = args.VideoFilter(videoFilter)
```

NodudeWasTaken replied:
I have some concerns, one is that pixel format isn't enforced, which may cause browser incompatibilities.
The reason hwupload and hwdownload was there was specifically to accommodate changing the format, but it might be the case that the encoder is nice enough to be consistent (NVENC isnt).
I also assume you mean to always apply the video filter, so i moved it out of the scale check (in stream.go).
https://github.com/stashapp/stash/blob/6295dc4d5d70e762959b4e1d88fc944f39f8b5cf/pkg/ffmpeg/stream.go#L300-L305

NodudeWasTaken replied:
I changed the intel and nvidia encoder to this format.
Given that i can't really test the vaapi encoder, i think it should be disabled for this commit, unless you want to work on it.
So it can be pushed ahead, and maybe someone with an actual amd gpu can make a commit fixing the amd encoders.

bnkai commented:
@NodudeWasTaken i have pasted the below `ffmpeg -hide_banner -vaapi_device /dev/dri/renderD128 -f lavfi -i testsrc=duration=5:size=320x240:rate=25 -vf format=nv12,hwupload,scale_vaapi=-2:720 -c:v h264_vaapi -movflags frag_keyframe+empty_moov -qp 20 -f mp4 test.mp4` command in discord as well to see if the issue i have is with my card or more generic (my card doesnt support vp8/vp9 hw encode so nothing else i can try ). If you need some ffmpeg command tested for the hw detection / encoding part it might be easier to post it here or in the  discord channel for some feedback?

NodudeWasTaken commented:
> @NodudeWasTaken i have pasted the below `ffmpeg -hide_banner -vaapi_device /dev/dri/renderD128 -f lavfi -i testsrc=duration=5:size=320x240:rate=25 -vf format=nv12,hwupload,scale_vaapi=-2:720 -c:v h264_vaapi -movflags frag_keyframe+empty_moov -qp 20 -f mp4 test.mp4` command in discord as well to see if the issue i have is with my card or more generic (my card doesnt support vp8/vp9 hw encode so nothing else i can try ). If you need some ffmpeg command tested for the hw detection / encoding part it might be easier to post it here or in the discord channel for some feedback?

The commands that need testing:

Intel VP9 (10th gen or above):
`ffmpeg -init_hw_device qsv=hw -filter_hw_device hw -f lavfi -i color=c=red -t 0.1 -c:v vp9_qsv -global_quality 25 -preset fast -vf hwupload=extra_hw_frames=64,format=qsv,scale_qsv=-1:240 intel_vp9.webm`

VAAPI H264:
I think this one is confirmed as broken
`ffmpeg -vaapi_device /dev/dri/renderD128 -f lavfi -i color=c=red -t 0.1 -c:v h264_vaapi -movflags frag_keyframe+empty_moov -qp 20 -vf format=nv12,hwupload,scale_vaapi=-2:240 vaapi_h264.mp4`

VAAPI VP9:
Might need `-bsf:v vp9_raw_reorder,vp9_superframe`
`ffmpeg -vaapi_device /dev/dri/renderD128 -f lavfi -i color=c=red -t 0.1 -c:v vp9_vaapi -qp 20 -vf format=nv12,hwupload,scale_vaapi=-2:240 vaapi_vp9.webm`

AMD AMF H264:
`ffmpeg -f lavfi -i color=c=red -t 0.1 -c:v h264_amf -movflags frag_keyframe+empty_moov -quality speed -pix_fmt yuv420p -vf scale=-2:240 amd_h264.mp4`

macOS H264:
`ffmpeg -f lavfi -i color=c=red -t 0.1 -c:v h264_videotoolbox -movflags frag_keyframe+empty_moov -prio_speed 1 -pix_fmt yuv420p -vf scale=-2:240 macos_h264.mp4`

OpenMAX H264:
`ffmpeg -f lavfi -i color=c=red -t 0.1 -c:v h264_omx -movflags frag_keyframe+empty_moov -preset superfast -crf 25 -pix_fmt yuv420p -vf scale=-2:240 openmax_h264.mp4`

DingDongSoLong4 commented on pkg/ffmpeg/stream_transcode.go:
```diff
@@ -64,8 +104,8 @@ var (
-		Args: func(videoFilter VideoFilter, videoOnly bool) (args Args) {
-			args = args.VideoCodec(VideoCodecCopy)
+		Args: func(codec VideoCodec, videoFilter VideoFilter, videoOnly bool) (args Args) {
+			args = CodecInit(codec)
```
For MKV the codec should stay as copy - that just copies the stream directly and does no transcoding at all, so using hardware acceleration would actually negatively impact performance.

DingDongSoLong4 commented on pkg/ffmpeg/stream_transcode.go:
```diff
@@ -83,28 +123,49 @@ var (
+	VideoFile     *file.VideoFile
+	Resolution    string
+	StartTime     float64
+	HardwareAccel bool
```
Here passing `HardwareAccel` through `TranscodeOptions` isn't necessary because the config manager is available on `StreamManager.config`. You just need to add `GetTranscodeHardwareAcceleration()` to `StreamManagerConfig` to be able to use it.
If you look at #3491 you'll see I added the `StreamManager` as an argument to `makeStreamArgs` so you'll be able to directly use it from there.

NodudeWasTaken replied:
It is done

JanJastrow commented:
Hi, this looks like a great feature.
Will/Can/Should this also be used for generating `Scene Scrubber Sprites`, `Previews`, `PHash` etc? Since they all use FFmpeg, too. (AFAIK)

WithoutPants commented on pkg/ffmpeg/codec_detector.go:
```diff
@@ -0,0 +1,240 @@
+}*/
+
+// Prepend input for hardware encoding only
+func HWDeviceInit(args Args, codec VideoCodec) Args {
```
Unless this function (and the other functions in this file) are used outside of the package, they should probably be made private.

WithoutPants commented on pkg/ffmpeg/codec_detector.go:
```diff
@@ -0,0 +1,240 @@
+package ffmpeg
```
This file would probably better named `codec_hardware.go`. I would move the hardware codec declarations into this file as well.

WithoutPants commented on docker/build/x86_64/Dockerfile-CUDA:
```diff
@@ -0,0 +1,53 @@
+# This dockerfile should be built with `make docker-build` from the stash root.
+
+# Build Frontend
+FROM node:alpine as frontend
+RUN apk add --no-cache make
+## cache node_modules separately
+COPY ./ui/v2.5/package.json ./ui/v2.5/yarn.lock /stash/ui/v2.5/
+WORKDIR /stash
+RUN yarn --cwd ui/v2.5 install --frozen-lockfile.
+COPY Makefile /stash/
+COPY ./graphql /stash/graphql/
+COPY ./ui /stash/ui/
+RUN make generate-frontend
+ARG GITHASH
+ARG STASH_VERSION
+RUN BUILD_DATE=$(date +"%Y-%m-%d %H:%M:%S") make ui
+
+# Build Backend
+FROM golang:1.19-bullseye as backend
+RUN apt update && apt install -y build-essential golang
+WORKDIR /stash
+COPY ./go* ./*.go Makefile gqlgen.yml .gqlgenc.yml /stash/
+COPY ./scripts /stash/scripts/
+COPY ./vendor /stash/vendor/
+COPY ./pkg /stash/pkg/
+COPY ./cmd /stash/cmd
+COPY ./internal /stash/internal
+COPY --from=frontend /stash /stash/
+RUN make generate-backend
+ARG GITHASH
+ARG STASH_VERSION
+RUN make build
```
This is duplicated from the `Dockerfile` in the same directory. Could we not make this the build docker image use `FROM stash/build` then add the stuff below?

NodudeWasTaken replied:
The main problem is that only the frontend can be shared.
The backend must be built on a glibc platform such that it can run on Ubuntu.
Where alpine uses musl and makes Ubuntu incompatible binaries.

WithoutPants replied:
Sorry didn't realise it needed a different platform. Can you revert https://github.com/stashapp/stash/pull/3419/commits/444badbd49a31311503e6a807cb5522c8d142ebe?

NodudeWasTaken replied:
> Sorry didn't realise it needed a different platform. Can you revert [444badb](https://github.com/stashapp/stash/commit/444badbd49a31311503e6a807cb5522c8d142ebe)?

Reverted it

NodudeWasTaken commented on pkg/ffmpeg/codec_detector.go:
```diff
@@ -0,0 +1,240 @@
+
+	err := cmd.Run()
+	return err == nil
+}*/
```
I removed unused functions

NodudeWasTaken commented on pkg/ffmpeg/codec_detector.go:
```diff
@@ -0,0 +1,240 @@
+var HWCodecSupport []VideoCodec
+
+// Tests all (given) hardware codec's
+func FindHWCodecs(ctx context.Context, encoder FFMpeg) {
```
I have addressed this, please check that its satisfactory

WithoutPants replied:
Given that `FindHWCodecs` requires a context, I think make `NewEncoder` accept only the ffmpeg path and don't do the `FindHWCodecs` initialisation. `FindHWCodecs` can be renamed to something like `InitHWSupport` or similar, and re-add the call to the `manager` code after initialisation. Apologies for not getting it right the first time.

NodudeWasTaken replied:
> Given that `FindHWCodecs` requires a context, I think make `NewEncoder` accept only the ffmpeg path and don't do the `FindHWCodecs` initialisation. `FindHWCodecs` can be renamed to something like `InitHWSupport` or similar, and re-add the call to the `manager` code after initialisation. Apologies for not getting it right the first time.

It is done

NodudeWasTaken commented:
> Hi, this looks like a great feature. Will/Can/Should this also be used for generating `Scene Scrubber Sprites`, `Previews`, `PHash` etc? Since they all use FFmpeg, too. (AFAIK)

As far as i can tell, only previews use hardware encoding/sprites/screenshots/phash (all lead to screenshot encoder) could arguably benefit from using the "ffmpeg extra input args".
The only thing is that previews have a configureable preset, meaning i would have to convert it for each codec.
I will ask in the discord is this is desireable.

WithoutPants commented:
> As far as i can tell, only previews use software encoding, sprites/screenshots/phash (all lead to screenshot encoder) could arguably benefit from using the "ffmpeg extra input args". My only issue is that previews have a configureable preset, meaning i would have to convert it for each codec. I will ask in the discord is this is desireable.

This can be addressed in a separate PR as necessary. Let's not add more scope to this PR.

WithoutPants commented:
Note that the build is failing due to a merge issue against `develop` branch. Could you merge or rebase against develop and fix the compile issues?

I think some extra logging is necessary for debugging and tracing purposes, and it should use loglevel of `error` so that we can output the resulting error if a support call fails. On my machine, I had no apparent hardware support and had to add some logging and run the commands manually to determine why.

In my local build I added `logger.Tracef("Running command: %s", cmd.String())` before running so that we can see the command being run in the trace log. See the `FFMpeg.Generate` code for how to output the error message - this should be logged at the Debug log level if the command fails.

In my personal case, I was getting an error message about the `-preset` value `p2` used in the `nvenc` test. This was due to my `ffmpeg` version being too old. As far as I can tell, the preset isn't used in the actual encoding, so either a different preset should be used for the test, or the preset should be removed. The error message I then got was related to the drivers. Once I updated my drivers, along with my updated ffmpeg, I finally got `h264_nvenc` to show as supported.

NodudeWasTaken commented:
> I think some extra logging is necessary for debugging and tracing purposes, and it should use loglevel of `error` so that we can output the resulting error if a support call fails. On my machine, I had no apparent hardware support and had to add some logging and run the commands manually to determine why.
> 
> In my local build I added `logger.Tracef("Running command: %s", cmd.String())` before running so that we can see the command being run in the trace log. See the `FFMpeg.Generate` code for how to output the error message - this should be logged at the Debug log level if the command fails.
> 
Alright, now it prints hopefully useful debug messages in the log

> In my personal case, I was getting an error message about the `-preset` value `p2` used in the `nvenc` test. This was due to my `ffmpeg` version being too old. As far as I can tell, the preset isn't used in the actual encoding, so either a different preset should be used for the test, or the preset should be removed. The error message I then got was related to the drivers. Once I updated my drivers, along with my updated ffmpeg, I finally got `h264_nvenc` to show as supported.

The main problem is that the "legacy" presets are removed in NVVC SDK 12, and have been deprecated since SDK 10 (https://docs.nvidia.com/video-codec-sdk/12.0/deprecation-notices/index.html)
The preset is used for setting the speed/quality for the encode, to ensure enough speed.
I have left it out now so it defaults to p4 encoding to solve backwards compatibility.

twist3dimages commented:
Can confirm it works on Windows. Not my Unraid server using Docker with Intel or Nvidia

yd82 commented:
> 

I have the same question, I tried to config hw transcoding for a week on my unraid server with no success. Hope someone can point the right direction on how to do it. I don't know how to build the CUDA as described on Unraid.

NodudeWasTaken commented:
> Can confirm it works on Windows. Not my Unraid server using Docker with Intel or Nvidia is there a separate repo I need to pull? like stash:development-nvidia or like stash:development-intel
> I have the same question, I tried to config hw transcoding for a week on my unraid server with no success. Hope someone can point the right direction on how to do it. I don't know how to build the CUDA as described on Unraid.

It works on the CUDA build, im unaware of any plans or code to publish the CUDA build to dockerhub.
See https://github.com/stashapp/stash/issues/305#issuecomment-1463619496 for how to use

disconnect5852 commented:
I see: `[InitHWSupport] Supported HW codecs:`
Added in my docker compose yml:
`    devices:
      - /dev/dri/renderD128:/dev/dri/renderD128
      - /dev/dri/card0:/dev/dri/card0`
also `sudo chmod -R 777 /dev/dri` on host
vainfo:
`libva info: VA-API version 1.7.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_7
libva error: /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so init failed
libva info: va_openDriver() returns 1
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/i965_drv_video.so
libva info: Found init function __vaDriverInit_1_6
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.7 (libva 2.6.0)
vainfo: Driver version: Intel i965 driver for Intel(R) Bay Trail - 2.4.0
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Simple            :	VAEntrypointEncSlice
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointEncSlice
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileH264StereoHigh         :	VAEntrypointVLD
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileNone                   :	VAEntrypointVideoProc
      VAProfileJPEGBaseline           :	VAEntrypointVLD`

DogmaDragon commented:
@TwentySeven28 https://docs.stashapp.cc/getting-started/installation/unraid/#nvidia-runtime

TwentySeven28 commented:
> @TwentySeven28 https://docs.stashapp.cc/getting-started/installation/unraid/#nvidia-runtime

Thanks @DogmaDragon. That's super helpful. Any idea if I can retrofit this for Intel QuickSync?

DogmaDragon commented:
@TwentySeven28 check <https://github.com/stashapp/stash/issues/305>, I saw people talking about it there. Also you can hop into Discord for more help, it's a bit out of my wheel house.
